test/functional/pdo_sqlsrv/MsCommon_mid-refactor.inc

```diff
@@ -0,0 +1,1833 @@
+
+    public function __construct($dataType, $colName = null, $options = null, $encType = "deterministic")
+    {
+        if (is_null($colName)) {
+            $this->colName = get_default_colname($dataType);
+        } else {
+            $this->colName = $colName;
+        }
+        $this->dataType = $dataType;
+        $this->encType = $encType;
+        $this->options = $options;
+    }
+    /**
+     * @return string column definition for creating a table
+     */
+    public function getColDefOps()
+    {
+        return getColDef($this->colName, $this->dataType, $this->options, $this->encType);
+    }
+}
+
+
+/**
+ * @param  string  $colName : column name
+ * @param  string  $dataType : a string that includes the size of the type if necessary (e.g., decimal(10,5))
+ * @param  string  $options : a string that is null by default (e.g. NOT NULL Identity (1,1) )
+ * @param  string  $encType : randomized or deterministic; default is deterministic
+ * @return string column definition for creating a table
+ */
+function getColDef($colName, $dataType, $options = null, $encType = "deterministic")
+{
+    $append = " ";
+
+    // an identity column is not encrypted because a select query with identity column as the where clause is often run and the user want to have to bind parameter every time
+    if (isColEncrypted() && stripos($options, "identity") === false) {
+        $cekName = getCekName();
+        if (stripos($dataType, "char") !== false) {
+            $append .= "COLLATE Latin1_General_BIN2 ";
+        }
+        $append .= sprintf("ENCRYPTED WITH (ENCRYPTION_TYPE = %s, ALGORITHM = 'AEAD_AES_256_CBC_HMAC_SHA_256', COLUMN_ENCRYPTION_KEY = $cekName) ", $encType);
+    }
+    $append .= $options;
+    $colDef = "[" . $colName . "] " . $dataType . $append;
+    return $colDef;
+}
+
+
+/**
+ * @return string default column name when a name is not provided in the ColumnMeta class
+ */
+function getDefaultColName($dataType)
+{
+    $colName = "c_" . str_replace(",", "_", str_replace("(", "_", $dataType));
+    $colName = rtrim($colName, ")");
+    return $colName;
+}
+
+
+/**
+ * Create a table
+ * @param  object  $conn : PDO connection object
+ * @param  string  $tbname : name of the table to be created
+ * @param  array  $columnMetaArr : array of key value pair with column name as key and datatype as value, or array of columnMeta objects, which contain metadata for one column
+ * @param  array  $inputs : an associative array column name and its value; value may be a literal value or a BindParamOp object
+ * @param  bool  $disableCE : flag for disabling column encryption even when keystore is NOT none
+ *                          for creating table with datatypes not support for encryption
+ * @return int reporting the number of rows affected (should always be 0 for creating table)
+ */
+function createTable($conn, $tbname, $columnMetaArr, $disableCE = false)
+{
```

Comment:
The main reason is to make it easier to invoke. When you say the logic is in more than one place, do you mean getColDef() is outside of the ColumnMeta class? But I think it makes sense to be a standalone function too, when only the column name and datatype are provided. I don't quite understand why it's hard to maintain. Can you give an example? Is it hard to maintain having to handle two types in createTable? or is it hard to maintain to have the getColDef() outside of ColumnMeta?

Reply:
Harder to maintain because the logic of creating a column definition is no longer encapsulated in a class but in more than one place. The helper method createTable also knows about how it works, so it's not a black box.

Reply:
I will refactor it so the function stays within the ColumnMeta class. In createTable, it'll create a ColumnMeta class from the key value pairs and then call getColDef